install_config/persistent_storage/persistent_storage_flex_volume.adoc

```diff
@@ -0,0 +1,304 @@
+<3> Name of the driver. This field is mandatory.
+<4> Optional filesystem that is present on the volume.
+<5> Optional reference to a secret. Keys and value from this secret will be
+provided to the Flex Volume driver on invocation.
+<6> Optional read-only flag.
+<7> Additional options for the Flex Volume driver. This is a free-form
+dictionary of parameters that will be provided to the Flex Volume driver and
+typically includes name of the volume that this `PersistentVolume` represents,
+addresses of the storage server(s) and such.
+====
+
+=== Flex Volume drivers
+
+As noted above, a Flex Volume driver is an executable, typically a shell script,
+that resides in a well defined directory on all machines in the cluster, both
+masters and nodes. {product-title} calls it whenever it needs to attach, detach,
+mount or unmount a volume represented by a `PersistentVolume` with `flexVolume`
+as source.
+
+The first command line argument of the driver is always an operation name.
+Other parameters are specific to each operation. Most of the operations expect
+JSON string as a parameter. This parameter is literally a whole JSON string,
+not name of a file with the JSON! It contains all `flexVolume.options`, some
+options from `flexVolume` prefixed by `kubernetes.io/` and content of the
+referenced secret (if specified) prefixed by `kubernetes.io/secret/`.
+
+.Flex Driver JSON input
+====
+[source,json]
+----
+{
+	"fooServer": "192.168.0.1:1234", <1>
+        "fooVolumeName": "bar",
+	"kubernetes.io/fsType": "ext4", <2>
+	"kubernetes.io/readwrite": "ro", <3>
+	"kubernetes.io/secret/<key name>": "<key value>", <4>
+	"kubernetes.io/secret/<another key name>": "<another key value>",
+}
+----
+<1> All options from `flexVolume.options`.
+<2> Value of `flexVolume.fsType`.
+<3> `ro`/`rw` based on `flexVolume.readOnly`.
+<4> All keys and their values from the secret referenced by `flexVolume.secretRef`.
+====
+
+OpenShift expects a JSON on standard output of the driver. When not specified
+otherwise, the output describes result of the operation.
+
+.Flex Driver default output
+====
+[source,json]
+----
+{
+	"status": "<Success/Failure/Not supported>",
+	"message": "<Reason for success/failure>"
+}
+----
+====
+
+Exit code of the driver should be 0 for success and 1 for error.
+
+Operations should be idempotent, i.e. attaching of already attached volume or
+mounting of already mounted volume should result in operation success.
+
+Flex Volume driver can work in two modes, with attach/detach operation or
+without it. Attach/detach operation is used by OpenShift master
+to attach a volume to a node and to detach it from a node. This
+is useful when a node becomes unresponsive from any reason - master then can
+kill all pods on the node, detach all volumes from it and attach the volumes to
+another nodes to resume the applications, while the original node is still not
+reachable.
+
+Not every storage backend supports master-initiated detachment of a volume from
```

Comment:
[IMPORTANT]
    ====
    Not every storage back-end supports master-initiated detachment of a volume from another machine. Carefully consider what kind of Flex Volume driver you are you going to write.
    ====